keycloak-plugin/sasl-plugin/src/main/java/io/enmasse/keycloak/spi/SaslAuthenticator.java

```diff
@@ -82,11 +84,41 @@ public void init(final NetSocket socket,
         this.sasl = transport.sasl();
         sasl.server();
         sasl.allowSkip(false);
-        sasl.setMechanisms(MECHANISMS.keySet().toArray(new String[MECHANISMS.size()]));
-        keycloakSession = keycloakSessionFactory.create();
+        sasl.setMechanisms(getValidMechanisms(getPasswordHashAlgorithms()));
         connection = protonConnection;
     }
 
+    private String[] getValidMechanisms(Set<String> hashAlgos) {
+        TreeSet<SaslServerMechanism> mechanisms = new TreeSet<>(Comparator.comparingInt(SaslServerMechanism::priority));
+        for(SaslServerMechanism mech : ALL_MECHANISMS) {
+            for(String hashAlgo : hashAlgos) {
+                if(mech.isSupported(hashAlgo)) {
+                    mechanisms.add(mech);
+                    break;
+                }
+            }
+        }
+        return mechanisms.stream().map(SaslServerMechanism::getName).toArray(String[]::new);
+    }
+
+    private Set<String> getPasswordHashAlgorithms() {
+        Set<String> hashAlgos = new HashSet<>();
+        KeycloakSession keycloakSession = keycloakSessionFactory.create();
+        KeycloakTransactionManager transactionManager = keycloakSession.getTransactionManager();
+        transactionManager.begin();
+        try {
+            List<RealmModel> realms = keycloakSession.realms().getRealms();
+            for(RealmModel realm : realms) {
+                if(realm.getAttribute("enmasse-realm",Boolean.FALSE)) {
```

Comment:
@rgodfrey In case of an external Keycloak being used as authentication service, this attribute check now adds the requirement of having the "enmasse-realm" property in the manually created realm.

In case this property hasn't been set (and `hashAlgos` is empty), would it make sense to let `getValidMechanisms` return at least the mechanisms, that don't really check the hashAlgo (i.e. PlainSaslServerMechanism, XOAUTH2SaslServerMechanism)?
If not, I think it would be good to log a WARN message that no realm with this property exists and therefore authentication will fail.

Reply:
@calohmn yeah, it would probably make sense to add such always available mechanisms, I'll raise an issue for that and put it in the next release.